bftengine/src/bftengine/SeqNumInfo.hpp

```diff
@@ -173,14 +186,79 @@ class SeqNumInfo {
     static IncomingMsgsStorage& incomingMsgsStorage(void* context);
   };
 
+  class ExFuncForFastPathOptimisticCollector {
+   public:
+    // external messages
+    static FullCommitProofMsg* createCombinedSignatureMsg(void* context,
+                                                          SeqNum seqNumber,
+                                                          ViewNum viewNumber,
+                                                          const char* const combinedSig,
+                                                          uint16_t combinedSigLen,
+                                                          const concordUtils::SpanContext& span_context);
+
+    // internal messages
+    static InternalMessage createInterCombinedSigFailed(SeqNum seqNumber,
+                                                        ViewNum viewNumber,
+                                                        const std::set<uint16_t>& replicasWithBadSigs);
+    static InternalMessage createInterCombinedSigSucceeded(SeqNum seqNumber,
+                                                           ViewNum viewNumber,
+                                                           const char* combinedSig,
+                                                           uint16_t combinedSigLen,
+                                                           const concordUtils::SpanContext& span_context);
+    static InternalMessage createInterVerifyCombinedSigResult(SeqNum seqNumber, ViewNum viewNumber, bool isValid);
+
+    // from the ReplicaImp object
+    static uint16_t numberOfRequiredSignatures(void* context);
+    static std::shared_ptr<IThresholdVerifier> thresholdVerifier(SeqNum seqNumber);
+    static concord::util::SimpleThreadPool& threadPool(void* context);
+    static IncomingMsgsStorage& incomingMsgsStorage(void* context);
+  };
+
+  class ExFuncForFastPathThresholdCollector {
```

Comment:
ExFuncForFastPathOptimisticCollector and ExFuncForFastPathThresholdCollector looks same. So is it possible to create a single class Collector and use to using  <type traits1> = Collector<templates for type trait1> and using <type traits2> = Collector<templates for type trait 2>
And use the type traits instead of separate classes. Please check if this is feasible?
Another idea is to use a single collector interface and use the interface instead if type traits.

Reply:
If it's too much of a design change then ignore the comment, we can do the cleanup later.